File: src/main/python/sorting/countingsort/counting_sort.py
from math import inf


def counting_sort(to_sort):
    if len(to_sort) == 0:
        return

    min_num, max_num = inf, -inf

    for i in range(len(to_sort)):
        min_num = min(to_sort[i], min_num)
        max_num = max(to_sort[i], max_num)

    count = [0] * (max_num - min_num + 1)

    for i in range(len(to_sort)):
        count[to_sort[i] - min_num] += 1

    k = 0

    for i in range(len(count)):
        while count[i] > 0:
            to_sort[k] = i + min_num
            k += 1
            count[i] -= 1


if __name__ == "__main__":
    arr = [1, 2, 3, 6, 4, 5]
    print(str(arr))
    counting_sort(arr)
    print(str(arr))

File: java/src/main/java/sorting/countingsort/CountingSort.java
package sorting.countingsort;

import java.util.Arrays;

public class CountingSort {

  public static void sort(int[] arr) {
    int[] minMax = minAndMax(arr);
    int[] count = new int[minMax[1] - minMax[0] + 1];

    for (int value : arr) {
      count[value - minMax[0]]++;
    }

    int k = 0;
    for (int i = 0; i < count.length; i++) {
      while (count[i] > 0) {
        arr[k++] = i + minMax[0];
        count[i]--;
      }
    }
  }

  private static int[] minAndMax(int[] arr) {
    int min = Integer.MAX_VALUE, max = Integer.MIN_VALUE;

    for (int value : arr) {
      if (value < min) {
        min = value;
      }

      if (value > max) {
        max = value;
      }
    }

    return new int[] {min, max};
  }

  public static void main(String[] args) {
    int[] arr = {1, 5, 3, 5, 3, 2, 4, 5, 2};
    System.out.println(Arrays.toString(arr));
    sort(arr);
    System.out.println(Arrays.toString(arr));
  }
}
